### PR TITLE
Add `InputFormatterResult` and `InputFormatterContext.ModelName`

### PR DESCRIPTION
**Pull request by dougbu**
- #2722
- make communication of errors from formatters to `BodyModelBinder` explicit

nits:
- improve some doc comments
- add another `BodyValidationIntegrationTests` test

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Abstractions/Formatters/IInputFormatter.cs [hunk 1]
```diff
@@ -6,25 +6,26 @@
...
     /// <summary>
-    /// Reads an object from the request body.
+    /// Reads an <c>object</c> from the request body.
     /// </summary>
     public interface IInputFormatter
     {
         /// <summary>
-        /// Determines whether this <see cref="IInputFormatter"/> can de-serialize
-        /// an object of the specified type.
+        /// Determines whether this <see cref="IInputFormatter"/> can deserialize an <c>object</c> of the
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Abstractions/Formatters/IInputFormatter.cs]: picking nits here, but this is referring to object as a generic programming term, not as the specific class :laughing:

File: src/Microsoft.AspNet.Mvc.Abstractions/Formatters/InputFormatterResult.cs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        /// <summary>
+        /// Returns an <see cref="InputFormatterResult"/> indicating the <see cref="IInputFormatter.ReadAsync"/>
+        /// operation was successful.
+        /// </summary>
+        /// <param name="model">The deserialized <see cref="object"/>.</param>
+        /// <returns>
+        /// An <see cref="InputFormatterResult"/> indicating the <see cref="IInputFormatter.ReadAsync"/>
+        /// operation succeeded i.e. with <see cref="HasError"/> <c>false</c>.
+        /// </returns>
+        public static InputFormatterResult Successful(object model)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Abstractions/Formatters/InputFormatterResult.cs]: Can we make this `Success` instead?

File: src/Microsoft.AspNet.Mvc.Core/Formatters/InputFormatter.cs [hunk 3]
```diff
@@ -67,42 +67,48 @@ public virtual bool CanRead(InputFormatterContext context)
...
 
         /// <summary>
-        /// Returns a value indicating whether or not the given type can be read by this serializer.
+        /// Determines whether this <see cref="InputFormatter"/> can deserialize an <c>object</c> of the given
+        /// <paramref name="type"/>.
         /// </summary>
-        /// <param name="type">The type of object that will be read.</param>
-        /// <returns><c>true</c> if the type can be read, otherwise <c>false</c>.</returns>
+        /// <param name="type">The <see cref="Type"/> of <c>object</c> that will be read.</param>
+        /// <returns><c>true</c> if the <paramref name="type"/> can be read; <c>false</c> otherwise.</returns>
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/InputFormatter.cs]: This is a more obtuse way to phrase it, why the change?

File: src/Microsoft.AspNet.Mvc.Core/Formatters/InputFormatter.cs [hunk 4]
```diff
@@ -67,42 +67,48 @@ public virtual bool CanRead(InputFormatterContext context)
...
 
         /// <summary>
-        /// Reads the request body.
+        /// Reads an <c>object</c> from the request body.
         /// </summary>
-        /// <param name="context">The <see cref="InputFormatterContext"/> associated with the call.</param>
-        /// <returns>A task which can read the request body.</returns>
-        public abstract Task<object> ReadRequestBodyAsync(InputFormatterContext context);
+        /// <param name="context">The <see cref="InputFormatterContext"/>.</param>
+        /// <returns>A task that on completion deserializes the request body.</returns>
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/InputFormatter.cs]: `<see cref="Task....` ?

File: src/Microsoft.AspNet.Mvc.Core/Formatters/InputFormatter.cs [hunk 5]
```diff
@@ -126,7 +132,12 @@ protected Encoding SelectCharacterEncoding(MediaTypeHeaderValue contentType)
             }
 
             // No supported encoding was found so there is no way for us to start reading.
-            throw new InvalidOperationException(Resources.FormatInputFormatterNoEncoding(GetType().FullName));
+            return null;
+        }
+
+        protected string GetNoEncodingMessage()
+        {
+            return Resources.FormatInputFormatterNoEncoding(GetType().FullName);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/InputFormatter.cs]: This is a wierd method to expose on the base class. I agree in general that throwing from `SelectCharacterEncoding` isn't ideal for extensibility - but why this change? What problem do you need to solve?

File: src/Microsoft.AspNet.Mvc.Formatters.Json/JsonInputFormatter.cs [hunk 6]
```diff
@@ -49,16 +49,21 @@ public JsonSerializerSettings SerializerSettings
...
 
         /// <inheritdoc />
-        public override Task<object> ReadRequestBodyAsync([NotNull] InputFormatterContext context)
+        public override Task<InputFormatterResult> ReadRequestBodyAsync([NotNull] InputFormatterContext context)
         {
             var type = context.ModelType;
             var request = context.HttpContext.Request;
-            MediaTypeHeaderValue requestContentType = null;
+
+            MediaTypeHeaderValue requestContentType;
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Formatters.Json/JsonInputFormatter.cs]: :dancer:

File: src/Microsoft.AspNet.Mvc.Core/Formatters/InputFormatter.cs (hunk 3 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/InputFormatter.cs]: The changes here are limited to the references e.g. "type" -> `<see cref="Type"/>` and `otherwise <c>false</c>` -> `<c>false</c> otherwise` i.e. consistent with the first phrase. What's obtuse now? In other words and with no judgement for you or anyone else, are you repeating your `<see cref="object"/>` nitpick?

File: src/Microsoft.AspNet.Mvc.Formatters.Json/JsonInputFormatter.cs [hunk 7]
```diff
@@ -66,31 +71,53 @@ public override Task<object> ReadRequestBodyAsync([NotNull] InputFormatterContex
...
                     var exception = e.ErrorContext.Error;
-                    context.ModelState.TryAddModelError(e.ErrorContext.Path, e.ErrorContext.Error);
+
+                    // Handle path combinations such as "" + "Property", "Parent" + "Property", or "Parent" + "[12]".
+                    var key = e.ErrorContext.Path;
+                    if (!string.IsNullOrEmpty(context.ModelName))
+                    {
+                        if (string.IsNullOrEmpty(e.ErrorContext.Path) || e.ErrorContext.Path[0] == '[')
+                        {
+                            key = context.ModelName + e.ErrorContext.Path;
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Formatters.Json/JsonInputFormatter.cs]: It's a little wierd to parse the `IsNullOrEmpty` case falling into here. I know it's a no-op, it just seems overly clever

File: src/Microsoft.AspNet.Mvc.Core/Formatters/InputFormatter.cs (hunk 5 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/InputFormatter.cs]: So you've now seen the problem I solved here? Let's discuss alternatives tomorrow.

File: test/Microsoft.AspNet.Mvc.Formatters.Json.Test/JsonInputFormatterTest.cs [hunk 8]
```diff
@@ -38,7 +38,11 @@ public void CanRead_ReturnsTrueForAnySupportedContentType(string requestContentT
             var contentBytes = Encoding.UTF8.GetBytes("content");
 
             var httpContext = GetHttpContext(contentBytes, contentType: requestContentType);
-            var formatterContext = new InputFormatterContext(httpContext, new ModelStateDictionary(), typeof(string));
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Formatters.Json.Test/JsonInputFormatterTest.cs]: Is there adequate coverage of the cases around `ModelName + Path` that crop up? I can't tell from the diff

File: src/Microsoft.AspNet.Mvc.Core/Formatters/InputFormatter.cs (hunk 5 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/InputFormatter.cs]: Yep. I'm sleepy but I realized you need this to not-throw so you can use the validation error instead. 

I actually feel a little bit like just living with duplication of the message is a better choice than the putting this turd here.

**Comment by rynowak** [comment]
:watch: this is much cleaner than I expected. Very nice!

File: src/Microsoft.AspNet.Mvc.Core/Formatters/InputFormatter.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/InputFormatter.cs]: Subtracting the markup:

> true if the type can be read, otherwise false.
> true if the type can be read; false otherwise.

A lot of these doc changes are improvements, but very tiny ones

File: src/Microsoft.AspNet.Mvc.Abstractions/Formatters/IInputFormatter.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Abstractions/Formatters/IInputFormatter.cs]: Agreed, that's why I used `<c>object</c>` instead of `<see cref="object"/>`. I can go back to "object" if you prefer. ??

File: src/Microsoft.AspNet.Mvc.Abstractions/Formatters/InputFormatterResult.cs (hunk 2 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Abstractions/Formatters/InputFormatterResult.cs]: Sure. And `Failure()` for the alternative -- to be consistent.

File: src/Microsoft.AspNet.Mvc.Core/Formatters/InputFormatter.cs (hunk 3 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/InputFormatter.cs]: I was changing some and chose to change all. Here, do you prefer the original comma and final word order?

File: src/Microsoft.AspNet.Mvc.Core/Formatters/InputFormatter.cs (hunk 4 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/InputFormatter.cs]: Should be. Thanx.

File: src/Microsoft.AspNet.Mvc.Formatters.Json/JsonInputFormatter.cs (hunk 7 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Formatters.Json/JsonInputFormatter.cs]: :ok: I'll add a block just for that case and do the `else if` thing.

File: test/Microsoft.AspNet.Mvc.Formatters.Json.Test/JsonInputFormatterTest.cs (hunk 8 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Formatters.Json.Test/JsonInputFormatterTest.cs]: Good catch. I meant to vary the new `FromBodyAndRequiredOnInnerValueTypeProperty_NotBound_JsonFormatterSuccessful()` until I got deserialization until it failed lower down but didn't go back and finish after rebasing. Will add another test or two here or in `BodyValidationIntegrationTests`...

File: src/Microsoft.AspNet.Mvc.Abstractions/Formatters/IInputFormatter.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Abstractions/Formatters/IInputFormatter.cs]: I'm not really sure what the right thing notationally is to do here - I just wanted to point out my concern. I trust whatever you think is the right way.

File: src/Microsoft.AspNet.Mvc.Core/Formatters/InputFormatter.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/InputFormatter.cs]: yes

File: src/Microsoft.AspNet.Mvc.Abstractions/Formatters/IInputFormatter.cs (hunk 1 above)
Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.Abstractions/Formatters/IInputFormatter.cs]: It should have no notation at all really. It's just a plain English word.

**Comment by dougbu** [comment]
updated

File: src/Microsoft.AspNet.Mvc.Core/Formatters/InputFormatter.cs [hunk 9]
```diff
@@ -126,7 +136,11 @@ protected Encoding SelectCharacterEncoding(MediaTypeHeaderValue contentType)
             }
 
             // No supported encoding was found so there is no way for us to start reading.
-            throw new InvalidOperationException(Resources.FormatInputFormatterNoEncoding(GetType().FullName));
+            context.ModelState.TryAddModelError(
+                context.ModelName,
+                Resources.FormatInputFormatterNoEncoding(GetType().FullName));
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/InputFormatter.cs]: yay

**Comment by rynowak** [comment]
:shipit:

**Comment by dougbu** [comment]
42017fa
